servers/rendering/renderer_rd/effects/ss_effects.cpp

```diff
@@ -447,45 +457,27 @@ void SSEffects::downsample_depth(RID p_depth_buffer, const Vector<RID> &p_depth_
 		}
 	}
 
+	RID shader = ss_effects.downsample_shader.version_get_shader(ss_effects.downsample_shader_version, downsample_mode);
 	int depth_index = use_half_size ? 1 : 0;
 
 	RD::get_singleton()->draw_command_begin_label("Downsample Depth");
-	if (p_invalidate_uniform_set || use_full_mips != ss_effects.used_full_mips_last_frame || use_half_size != ss_effects.used_half_size_last_frame || use_mips != ss_effects.used_mips_last_frame) {
-		if (ss_effects.downsample_uniform_set.is_valid() && RD::get_singleton()->uniform_set_is_valid(ss_effects.downsample_uniform_set)) {
-			RD::get_singleton()->free(ss_effects.downsample_uniform_set);
-			ss_effects.downsample_uniform_set = RID();
-		}
 
-		Vector<RD::Uniform> uniforms;
-		{
-			RD::Uniform u;
-			u.uniform_type = RD::UNIFORM_TYPE_IMAGE;
-			u.binding = 0;
-			u.append_id(p_depth_mipmaps[depth_index + 1]);
-			uniforms.push_back(u);
-		}
-		{
-			RD::Uniform u;
-			u.uniform_type = RD::UNIFORM_TYPE_IMAGE;
-			u.binding = 1;
-			u.append_id(p_depth_mipmaps[depth_index + 2]);
-			uniforms.push_back(u);
-		}
-		{
-			RD::Uniform u;
-			u.uniform_type = RD::UNIFORM_TYPE_IMAGE;
-			u.binding = 2;
-			u.append_id(p_depth_mipmaps[depth_index + 3]);
-			uniforms.push_back(u);
-		}
-		if (use_full_mips) {
-			RD::Uniform u;
-			u.uniform_type = RD::UNIFORM_TYPE_IMAGE;
-			u.binding = 3;
-			u.append_id(p_depth_mipmaps[4]);
-			uniforms.push_back(u);
+	RID downsample_uniform_set;
+	if (use_mips) {
+		// Grab our downsample uniform set from cache, these are automatically cleaned up if the depth textures are cleared.
+		// This also ensures we can switch between left eye and right eye uniform sets without recreating the uniform twice a frame.
+		RD::Uniform u_depth[4];
+
+		for (int i = 0; i < 4; i++) {
+			RID depth_mipmap = p_render_buffers->get_texture_slice(RB_SCOPE_SSDS, RB_LINEAR_DEPTH, p_view * 4, depth_index + i + 1, 4, 1);
+
+			u_depth[i].uniform_type = RD::UNIFORM_TYPE_IMAGE;
+			u_depth[i].binding = i;
+			u_depth[i].append_id(depth_mipmap);
```

Comment:
This is going to allocate more than we need when ``use_full_mips`` is false. That may not be a big deal

Reply:
Well that part of the code is the same as before, it just used to be written out.

I believe it either uses mips, or it doesn't, and it skips this code if it doesn't

Reply:
Had another look at this and saw what I missed in the original code, corrected this and added some comments to clarify how this works.